### PR TITLE
enh(cpp) simpler struct matching

### LINKED ISSUE #2520: Add some type of look-behind support
**Is your request related to a specific problem you're having?**

Yes, there is no easy way to do look behind or to say: 

> match this, but only if immediately proceeded by that...

It's possible, but it gets ugly fast. For example here's how to match say `instanceof [class]` and do it _with no ambiguity_: (where [class] is the only part I want highlighted)

```js
      {
        begin: regex.concat('(?=', /\binstanceof /, IDENT_RE, ')'),
        contains: [
          { begin: /\binstanceof /,
            keywords: KEYWORDS,
            starts: {
              contains: [
                { begin: IDENT_RE, className: "title"}
              ],
            }
          },
        ],
      },
```

First we match the whole thing with a look-ahead... then our mode starts and our first rule has to re-include keywords from above (because now the context has changed), and then we have to match `instance of` and then start a new mode [for `IDENT_RE`] with `starts` to avoid the chance of our two expressions being similar enough that the first rule might fire for the second.

**The solution you'd prefer / feature you'd like to see added...**

Well if JS just supported look behind, we'd be all set but since it doesn't... Sample code from Javascript we're trying to match:

```js
if (str instanceof Token) { ...
// and
var token = new Token()
```

And how we might go about doing it:

```js
 { 
  skipBefore: /(\binstanceof[ ]+)/,
  begin: IDENT_RE,
  className: "title"
}, {
  skipBefore: /(\bnew[ ]+)/,
  begin: IDENT_RE,
  className: "title" 
},
// and for comparison: look-ahead, which is already possible;
{
  begin: regex.concat(IDENT_RE, /(?=\.(constructor|prototype))/),
  className: "title"
},
```

This is actually only a 5 line patch. Building this into the parser means the mode context doesn't change, so your existing keyword scope "just works"... and these become very simple rules.

**Potential Caveats**

1. It might have interactions with `returnBegin`... I'm not sure why you'd ever want to use them both... but we could either:
   - have `returnBegin` disable skipBefores "skip" ability, making it more of a "prefix"
   - throw an error for grammar developers and just say these two things don't mix.

2. The naming isn't perfect... technically this is `matchBeforeAndSkip`... I hesitate to name it `lookbehind` though because that's NOT what it's doing... it's still looking forward the whole time.  If another rule has consumed the content already this can't match it by "looking backwards", because there is no backwards looking going on.

**Any alternative solutions you considered...**

Writing code to write the complex version... but the keyword complexity and mode context switches make it a pain even if you had a helper, since you'd still have to pass in KEYWORDS from the parent context (as far as I can see)...

### LINKED ISSUE #2736: (c) Declaration of struct variable triggers class title mode
**Describe the issue**

When declaring a variable of type `struct foobar`, this triggers the rule for a struct/class title, which leads to the whole declaration being highlighted as a title. See https://meta.stackexchange.com/a/354753/1968.

**Which language seems to have the issue?**

`c`/`cpp`

**Are you using `highlight` or `highlightAuto`?**

`highlight`

**Sample Code to Reproduce**

```c
struct List *newnode = (struct List *)malloc(size * sizeof(struct List));
```

This is rendered as

```xml
<class><keyword>struct</keyword> <title>List</title> *<title>newnode</title> = (<title>struct</title> <title>List</span> *)<span class="hljs-title">malloc</title>(<title>size</title> * <title>sizeof</title>(<title>struct</title> <title>List</title>));</class>
```

**Expected behavior**

Should render as

```xml
<keyword>struct</keyword> List *newnode = (<keyword>struct</keyword> List *)<built_in>malloc</built_in>( * <keyword>sizeof</keyword>(<keyword>struct</keyword> List));
```

**Additional context**

Without the initial `struct` the highlighter doesn’t trigger class/title mode. It still doesn’t correctly render the `struct` keywords in the rest of the initialisation but that’s already addressed (incidentally) by #2728.

### LINKED ISSUE #3112: Highlight identifier after 'impl' keyword in Rust language definition
Hi! 👋 
      
Firstly, thanks for your work on this project! 🙂

Today I used [patch-package](https://github.com/ds300/patch-package) to patch `highlight.js@9.18.3` for the project I'm working on.

It looks as if the identifier after the `impl` keyword is not being highlighted. I've added an extra element to the class definition to include `impl`.

Here is the diff that solved my problem:

```diff
diff --git a/node_modules/highlight.js/lib/languages/rust.js b/node_modules/highlight.js/lib/languages/rust.js
index 5e2af44..590dc15 100644
--- a/node_modules/highlight.js/lib/languages/rust.js
+++ b/node_modules/highlight.js/lib/languages/rust.js
@@ -89,7 +89,7 @@ module.exports = function(hljs) {
       },
       {
         className: 'class',
-        beginKeywords: 'trait enum struct union', end: '{',
+        beginKeywords: 'trait enum struct union impl', end: '{',
         contains: [
           hljs.inherit(hljs.UNDERSCORE_TITLE_MODE, {endsParent: true})
         ],
```

<em>This issue body was [partially generated by patch-package](https://github.com/ds300/patch-package/issues/296).</em>

### PR DESCRIPTION
**Pull request by joshgoebel**
Resolves #2736.   Resolves #2520. Resolves #3112.

### Changes

- Adds new `beforeMatch` compiler extension
- (cpp) Simplified struct/class/enum/union matcher
- (java) Simplified class interface enum matcher
- adds `title.class` sub-scope support. Related #2521.

Seems to do about as well as GitHub now.

Screen Shot 2021-03-26 at 12 39 58 PM

### Checklist
- [x] Added markup tests, or they don't apply here because...
- [x] docs for `beforeMatch`
- [x] Updated the changelog at `CHANGES.md`

### CODE REVIEW

**Comment by joshgoebel** [comment]
Right now I'm using .title-class... but I wonder if perhaps we shouldn't a bit grander:

.title--class
.title__class

**Comment by joshgoebel** [comment]
This is currently rebased on the multi-class stuff (so I could finish it), but in review that can be ignored.

**Review by joshgoebel — COMMENTED** [review]
Self-approving.  @allejo If you have time to look at this later and see anything terrible please let me know.

File: .eslintrc.js [hunk 1]
```diff
@@ -12,7 +12,7 @@ module.exports = {
     ecmaVersion: 2015,
     sourceType: "module"
   },
-  parser: '@typescript-eslint/parser',
+  // parser: '@typescript-eslint/parser',
```
Comment (NullVoxPopuli) [inline .eslintrc.js:15]: why this change?

Comment (joshgoebel) [reply .eslintrc.js:15]: It goes back and forth. :)  The uncomment never should have gotten committed.

Comment (joshgoebel) [reply .eslintrc.js:15]: Do you know if you can use `parser` only for specific paths?

File: src/highlight.js [hunk 2]
```diff
@@ -452,7 +453,7 @@ const HLJS = function(hljs) {
         modeBuffer += codeToHighlight.slice(match.index, match.index + 1);
         if (!SAFE_MODE) {
           /** @type {AnnotatedError} */
-          const err = new Error('0 width match regex');
+          const err = new Error(`0 width match regex (${languageName})`);
```
Comment (NullVoxPopuli) [inline src/highlight.js:456]: I appreciate the specification!

File: src/languages/actionscript.js [hunk 3]
```diff
@@ -92,20 +96,16 @@ export default function(hljs) {
       hljs.C_BLOCK_COMMENT_MODE,
       hljs.C_NUMBER_MODE,
       {
-        className: 'class',
-        beginKeywords: 'package',
-        end: /\{/,
-        contains: [ hljs.TITLE_MODE ]
+        beforeMatch: /\b(package)\s+/,
```
Comment (NullVoxPopuli) [inline src/languages/actionscript.js:99]: this is super slick!

**Review by NullVoxPopuli — APPROVED** [review]
This is exciting! It'll make my own grammar cleaner in a few places :D

File: src/languages/actionscript.js (hunk 3 above)
Comment (joshgoebel) [reply src/languages/actionscript.js:99]: It may still be replaced by multi-match though since that can do the same thing... not sure.

File: .eslintrc.js (hunk 1 above)
Comment (NullVoxPopuli) [reply .eslintrc.js:15]: you can! via the overrides options (which I recommend moving to _only_ overrides, as I've done here: https://github.com/NullVoxPopuli/eslint-configs)

Comment (joshgoebel) [reply .eslintrc.js]: Nice!  For some reason I was thinking that something as high level as the parser couldn't be configured in overrides. Seems to work just fine though.
